### PR TITLE
fix(spur-cloud-api): adapt GPU allocation to CDI-based device API

### PR DESCRIPTION
**Pull request by shiv-tyagi**
## Summary

- Adapts `get_gpu_capacity` to the new CDI-based device API introduced in spur PR #262. The flat `alloc.gpus` field was replaced upstream with a generic `alloc.devices` map keyed by device class — this reads allocated GPUs from `alloc.devices["gpu"]` and cross-references device IDs against total resources for type resolution.
- Bumps `tonic` 0.12→0.14 and `prost` 0.13→0.14 to match the updated `spur-proto` dependency.

## Test plan

- [x] `cargo build` — compiles cleanly against latest spur-proto from main
- [x] Deployed on K8s test cluster with fake CDI GPU resources (3 nodes × 4 MI300X GPUs)
- [x] `/api/gpus` endpoint returns correct capacity, allocation counts, and per-node breakdown
- [x] Frontend loads and displays GPU dashboard

### CODE REVIEW

File: crates/spur-cloud-api/src/spur_client.rs [hunk 1]
```diff
@@ -434,18 +434,22 @@ pub async fn get_gpu_capacity(
...
+                                pool.allocated += dev.count as u32;
+                            }
+                        } else {
+                            warn!(
+                                node = %node.name,
+                                gpu_type = %gpu.gpu_type,
+                                "allocated GPU type not found in node's total resources - data inconsistency"
+                            );
+                        }
                     }
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply crates/spur-cloud-api/src/spur_client.rs]: Done. Added a `warn!` with node name and device_id when the lookup fails.

File: crates/spur-cloud-api/src/spur_client.rs [hunk 2]
```diff
@@ -459,8 +463,14 @@ pub async fn get_gpu_capacity(
...
             if let Some(alloc) = alloc_res {
-                for gpu in &alloc.gpus {
-                    *alloc_counts.entry(gpu.gpu_type.clone()).or_insert(0) += 1;
+                if let Some(devs) = alloc.devices.get("gpu") {
+                    for dev in &devs.devices {
+                        if let Some(gpu) = total.gpus.iter().find(|g| g.device_id == dev.device_id)
+                        {
+                            *alloc_counts.entry(gpu.gpu_type.clone()).or_insert(0) +=
+                                dev.count as u32;
+                        }
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply crates/spur-cloud-api/src/spur_client.rs]: Fixed in the same push. Both paths now log a warning on unresolved device_id.

File: Cargo.toml [hunk 3]
```diff
@@ -73,7 +73,7 @@ sha2 = "0.10"
 semver = "1"
 
 # Spur proto — pinned to ROCm/spur release tag (see Cargo.lock for exact commit)
-spur-proto = { git = "https://github.com/ROCm/spur.git", tag = "v0.3.0" }
+spur-proto = { git = "https://github.com/ROCm/spur.git", rev = "5781872" }
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply Cargo.toml]: Good catch. Updated the comment and switched to the full 40-char commit hash.

File: crates/spur-cloud-api/src/spur_client.rs [hunk 4]
```diff
@@ -434,18 +434,22 @@ pub async fn get_gpu_capacity(
...
-                if let Some(pool) = pools.get_mut(&gpu.gpu_type) {
-                    if schedulable {
-                        pool.allocated += 1;
+        if let (Some(alloc), Some(total)) = (alloc_res, total_res) {
+            if let Some(devs) = alloc.devices.get("gpu") {
+                for dev in &devs.devices {
+                    if let Some(gpu) = total.gpus.iter().find(|g| g.device_id == dev.device_id) {
+                        if let Some(pool) = pools.get_mut(&gpu.gpu_type) {
+                            if schedulable {
+                                pool.allocated += dev.count as u32;
```
[the remark this answers is not in the document]
Comment (shiv-tyagi) [reply crates/spur-cloud-api/src/spur_client.rs]: Added 6 tests covering: basic idle nodes, partial allocations, non-schedulable exclusion, multiple GPU types, count > 1, and unresolved device_id. Extracted the logic into a pure `compute_gpu_pools` function to make it testable without a gRPC client.

**Review by yansun1996 — COMMENTED** [review]
## Code Review

Refactor and test coverage are a clear improvement — extracting `compute_gpu_pools(&[NodeInfo])` makes the logic unit-testable without a live gRPC client, and the new cases cover idle/allocated/down/multi-type/unresolved-id paths well. A few items below, one of which may be a correctness bug.

### Top issue — verify `count` / `device_id` semantics (possible correctness bug)

The matching is:
```rust
total.gpus.iter().find(|g| g.device_id == dev.device_id)
...
pool.allocated += dev.count as u32;
```
This only makes sense if `device_id` is a **physical GPU id** in *both* `total.gpus` and `AllocatedDevice`. But if that's true, `count` for a single physical device should always be `1` — yet the test `gpu_capacity_count_greater_than_one` models `device_id: 0, count: 2` on a 2-GPU node and asserts `allocated == 2`, which reports device_id 1 as allocated when it's actually idle.

So one of two things is wrong:
- If `count` can legitimately be >1, then `device_id` is *not* a physical id (it's a class/model index), and matching it against `total.gpus[].device_id` mixes two id spaces — `available = total − allocated` becomes unreliable.
- If `device_id` *is* physical, then `count` is always 1 and the `count`-based math (and that test) models an impossible state.

Please confirm against the spur #262 proto what `AllocatedDevice.device_id` and `.count` actually mean, and make the test reflect a real CDI payload. This is the one item that could silently produce wrong dashboard numbers.

### Other notes
- **Double warning:** an unresolved `device_id` triggers `warn!` in the global block *and* again in the per-node block for the same device. Consider resolving once.
- **Silent drop when total is missing:** the global block is now `if let (Some(alloc), Some(total))`. A node with allocations but no `total_resources` now drops those allocations with no warning.
- **`dev.count as u32`:** if the proto field is `i32`, a negative value wraps to a huge `u32`. `u32::try_from` (or `.max(0)`) would be safer at the boundary.
- **Dependency tree:** tonic 0.14 pulls in `axum 0.8` while the app still uses `axum 0.7`, so two axum trees now compile. Functionally fine (tonic's axum is internal), just a larger build — worth noting.
- **`spur-proto` pinned to a raw commit** instead of a release tag. The PR body acknowledges this; please re-pin to a tag once spur cuts a release including #262, for reproducibility.

### Test gaps (once `count` is settled)
1. A multi-node + multi-allocation case confirming global vs per-node tallies agree.
2. The "alloc present, total absent" edge case noted above.

### Verdict
Looks good **pending** clarification of the `count`/`device_id` semantics — the only item that could affect correctness. Everything else is minor cleanup; the dep bumps and refactor are sound.

**Comment by shiv-tyagi** [comment]
Thanks for the thorough review.

**count/device_id semantics:** `device_id` is a physical GPU index and `count` is always 1 for GPUs today (spur-core uses `AllocatedDevice::injectable(id)` which hardcodes `count: 1`). The `count > 1` path only fires for countable generics like bandwidth/licenses. That said, we intentionally keep the `count`-based math in spur-cloud rather than hardcoding `count == 1`, so we don't break if spur later adds time-slicing or MIG support. Updated the test comment to make this explicit.

For the other items:

- **Double warning:** Refactored to resolve `device_id` through a per-node HashMap lookup in a single pass. Both global pool and per-node tallies now use the same resolved data.
- **Silent drop (alloc without total):** Now emits a warning and skips the node.
- **`dev.count as u32`:** Changed to `u32::try_from(...).unwrap_or(u32::MAX)`. Also switched pool accumulation to `saturating_add`.
- **Dual axum trees:** Noted, nothing actionable for now.
- **spur-proto pinned to commit:** Will re-pin to a tag once spur cuts a release with #262.

Added two new tests: alloc-without-total edge case, and a multi-node scenario that asserts global totals equal the sum of per-node tallies.

**Review by sgopinath1 — APPROVED** [review]
